core/src/main/java/org/elasticsearch/cluster/routing/allocation/FailedRerouteAllocation.java

```diff
@@ -33,14 +33,36 @@
  */
 public class FailedRerouteAllocation extends RoutingAllocation {
 
-    private final List<ShardRouting> failedShards;
+    /**
+     * A failed shard with the shard routing itself and an optional
+     * details on why it failed.
+     */
+    public static class FailedShard {
+        private final ShardRouting shard;
```

Comment:
I think we can make the members here `public final` and remove the getters then it's clear it's a struct like class?